### DISCUSSION TITLE
Support for GekiChuMai + WACCA

### DISCUSSION THREAD

**Discussion by albshin** [noahm/DDRCardDraw#116] [General]
Starting this thread to discuss whether you would be open to integrating Sega rhythm game song data into the application. There are a couple of differences between DDR/Konami games that may pose a few difficulties:

- GekiChuMai uses a (+) level system starting from level 7. For example, level 7 through 9 includes: 7, 7+, 8, 8+, 9+, 10
  - WACCA supports a similar level schema that goes up to 15 but skips 14+
- Songs use an internal level rating constant in addition to an (inconsistent) + level system
- maimai has two types of charts, STANDARD and DX. It is important to denote which type of chart a song is.
- BPM is fairly unimportant as the games use a CMOD-esque system
- Unlike DDR/SDVX, filtering by song categories is a bit more important

I've done a test run of importing maimai DX Universe data into my fork of the app: https://sparkling-chaja-ee9654.netlify.app/#game-maimai_universe

Some things that I've done to support this:
- Since the card app does not support + levels normally, I've denoted + levels by adding a .5 to the end of the level (for example 7+ is a 7.5 on the `lvl` field). In the places where levels need to be shown, a util function `formatLevel` is used to properly show the + levels
- The upper/lower level number inputs hasvebeen replaced with Select boxes
- Modified card draw logic that uses `times()` and replaced it with a function to aggregate a list of valid levels. Bucketing logic using `times()` assumed levels went up incrementally by 1
- Denoted the song's category and chart type as a flag. This works fine but does not look as nice in the UI.
- Manually added a maimai chart type indicator to the cards similar to how the "SHOCK" chart indicator works

As for other changes that need to happen:
- Internal level filtering will most likely require a new field that is not currently in the chart schema in addition to a set of Number inputs to support filtering

Given the challenges and potential for bloat, is it worthwhile to pursue adding support for these games or would maintaining a separate branch be more appropriate?

**Comment by albshin** [comment]
After sleeping on it, I think it would be best to maintain SEGA game support as a separate fork given all of the extra features needed (+ these games are not wildly available in the US). Closing

**Comment by noahm** [comment]
Hey, sorry I didn't see this quite as soon as I would have liked! I have given some thought to how games with the WACCA-style "plus difficulty" levels would fit into the current system, and I'm not opposed to doing some work to get them better support in the main repo here.

I agree it's easier to just keep those in a separate fork, and you're welcome to do so of course. That said, there are other benefits of getting it fully integrated here; and that's especially the case when it comes to saving you the work of merging in any major changes I might make here in the future (e.g. the `tournament-mode` branch). 

Your approach of providing a `formatLevel` utility to display plus difficulty is correct, IMO. However I would probably continue to represent the levels as integers in the game data to avoid more complex changes to the random draw logic (e.g. displayLvl = rawLvl / 2). Using flags more widely to categorize and filter chart types also checks out, and there can also be game specific options to help make the settings UI look and work better for the way flags are used within that one game's data. 

In summary, I'm open to exploring the ideas further in code, even if that ultimately leads me to the same conclusion you reached after a good night's sleep. 😛 (and of course I'm happy to answer questions and lend a hand with the more difficult merges if you do just stick with an ongoing fork)
